### DISCUSSION TITLE
HostRule via Docker env not recognized

### DISCUSSION THREAD

**Discussion by gbrehmer** [renovatebot/renovate#9010] [Request Help]
We are using a private Artifatory as Maven Repository which needs basic auth.

I tried to pass the host rule as env for docker container but it seems to be ignored (inside gitlab ci job = $xxx will be evaluated to secret variable):

```
docker run -e RENOVATE_HOST_RULES='[{"domainName":"artifactory.host.de","username":"$ARTIFACTORY_READ_USER","password":"$ARTIFACTORY_READ_TOKEN"}]' ...
```

```
docker run -e RENOVATE_HOST_RULES_ARTIFACTORY_HOSTNAME=artifactory.host.de -e RENOVATE_HOST_RULES_ARTIFACTORY_USERNAME=$ARTIFACTORY_READ_USER -e RENOVATE_HOST_RULES_ARTIFACTORY_PASSWORD=$ARTIFACTORY_READ_TOKEN ...
```

Result in both cases:
```
Dependency lookup unauthorized. Please add authentication with a hostRule (repository=xxx/xxx)
       "failedUrl": "https://artifactory.host.de/artifactory/..."
```

Do I have to add the settings to the renovate.json inside source repository? I couldn't find the documentation for passing private key as env varibale for the renovate docker container (self hosting) because then I have to encrypt the credentials

**Comment by rarkins** [comment]
It seems like it could be a bash problem. IIRC bash won't expand variables inside single quotation marks `'` but will inside double quotation marks `"`

**Reply by gbrehmer** [reply]
the variables are expanded directly by gitlab ci (before the execution) doc: https://docs.gitlab.com/ee/ci/variables/

**Reply by gbrehmer** [reply]
are any special settings in the gradle file important?
on dev machine (w/o docker, renovate etc.) we are able to pass the credentials with custom system env or within user specifc gradle.properties. Is it possible to pass env variables from renovate docker to the internally running gradle process?

our gradle file:
```
...
def artifactoryUser = System.getenv("ARTIFACTORY_USER")
def artifactoryPassword = System.getenv("ARTIFACTORY_PASSWORD")

repositories {
    maven { url 'https://repo.spring.io/milestone' }
    maven { url 'https://repository.jboss.org/nexus/content/repositories/releases' }
    maven { url 'https://oss.sonatype.org/content/repositories/releases' }
    maven { url 'https://repo.maven.apache.org/maven2' }
    maven { url 'https://oss.sonatype.org/content/repositories/jetty-releases/' }
    maven {
        url 'https://artifactory.host.de/artifactory/project'
        credentials {
            username = artifactoryUser != null ? artifactoryUser : project.property('artifactory_user')
            password = artifactoryPassword != null ? artifactoryPassword : project.property('artifactory_password')
        }
    }
}
...
```

**Reply by gbrehmer** [reply]
and another hint: artifactory creates app tokens which can be used as password for basic auth. It's a JWT. e.g. your encrypt homepage is not able to encrypt the string (result: false)

**Reply by rarkins** [reply]
Set trustLevel=high

**Reply by gbrehmer** [reply]
Is already set on boath placrs (docker run params & renovate.json):

Docker (syntax is gitlab ci specific, :
```docker run -e LOG_LEVEL=debug -e GITHUB_COM_TOKEN=$GITHUB_COM_TOKEN -e ARTIFACTORY_USER=$ARTIFACTORY_READ_USER -e ARTIFACTORY_PASSWORD=$ARTIFACTORY_READ_TOKEN -e RENOVATE_HOST_RULES='[{"domainName":"artifactory.inovex.de","username":"$ARTIFACTORY_READ_USER","password":"$ARTIFACTORY_READ_TOKEN"}]' renovate/renovate:latest --token=$GitToken --platform=gitlab --trust-level=high --endpoint=$ENDPOINT $REPOSITORIES```


Example json:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "docker": {
    "enabled": false
  },
  "packageRules": [
    {
      "packagePatterns": [
        ".*"
      ],
      "groupName": "dependencies"
    }
  ],
  "labels": [
    "auto-generated"
  ],
  "trustLevel": "high"
 }

```

**Reply by rarkins** [reply]
I still think the problem may be related to the single quotes (`'`) you wrap this env in:
 `RENOVATE_HOST_RULES='[{"domainName":"artifactory.inovex.de","username":"$ARTIFACTORY_READ_USER","password":"$ARTIFACTORY_READ_TOKEN"}]'`. 

Please google "bash variable substitution single quotes" to read about variable substitution for single quotes and how to work around it.

**Reply by gbrehmer** [reply]
and the other approach with single env variables (based on https://github.com/renovatebot/renovatebot.github.io/issues/26#issuecomment-643262599): 

```
docker run \
  -e RENOVATE_HOST_RULES_ARTIFACTORY_HOSTNAME=artifactory.host.de \
  -e RENOVATE_HOST_RULES_ARTIFACTORY_USERNAME=$ARTIFACTORY_READ_USER  \
  -e RENOVATE_HOST_RULES_ARTIFACTORY_PASSWORD=$ARTIFACTORY_READ_TOKEN \ 
  ...
```

**Reply by rarkins** [reply]
> and the other approach with single env variables (based on [renovatebot/renovatebot.github.io#26 (comment)](https://github.com/renovatebot/renovatebot.github.io/issues/26#issuecomment-643262599)):

I think that's just someone giving ideas about how it could work. It's not implemented/supported

**Reply by gbrehmer** [reply]
ah sorry I missunderstood that comment. Thanks for clarifing. And I will test with ```"'"$var"'"```Thank you, Feedback comes in 30minutes ;)

**Reply by gbrehmer** [reply]
It's working now 🥳  (I only have problems with gradle dynamic versioning '1.2.+' but probably its not supported...we can switch do fixed versioning)
Your first answer was the correct one 👍  I had several misunderstandings and sorry for the back-and-forth

EDIT:
And I didn't checked the debug log properly, because hostrules are logged out, where the problem/cause is visible too:
```
DEBUG: Env config
       "config": {
         "hostRules": [
           {
             "domainName": "artifactory.host.de",
             "username": "$ARTIFACTORY_READ_USER",
             "password": "***********"
           },
           {"hostType": "github", "domainName": "github.com", "token": "***********"}
         ]
       }
```
